Add cli arg for output file

Kristina-Pianykh opened the pull request:
Closes https://github.com/gopherguides/hype/issues/34

This change should not affect the existing functionality or break any interfaces where the output of the export operation is written to stdout by default.

## Changes

- implement support for `-o` flag for output file in the `cli` package
- add unit tests in the `cli` package
- add integration tests in the `hype_test` package (introduced in #31 , I renamed the file `cli_errors_test.go` to `cli_integration_test.go` to keep the integration tests together in one file)
- add rule in Makefile for local testing
- replace redirection operator `>` in the GH workflow with `-o` to prevent committing an empty README.md on error

## Test

```go
go test -v -race -count=1 cli_integration_test.go
```

This is a bit of an awkward implementation but I tried to work around the existing package interfaces not to break anything.
@corylanou would you mind taking a look? I'm sure some of it can be done better and I'll be happy to reiterate on my solution.

corylanou reviewed (changes requested):
This is really great!  Thanks for all your work on this.  I have just a couple comments/changes.  Happy to discuss as well.

Kristina-Pianykh commented on cmd/hype/cli/export.go:
```diff
@@ -191,9 +201,33 @@ func (cmd *Export) main(ctx context.Context, pwd string, args []string) error {
 		return err
 	}
 
+	if !cmd.OutPath.Exists() {
+		fmt.Fprint(os.Stdout, stdoutBuffer.String())
+	} else {
+		path := cmd.OutPath.Value()
+		file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+		if err != nil {
+			return err
+		}
+		fmt.Fprint(file, stdoutBuffer.String())
+		file.Close()
+	}
```
First of all, thank you for the review and fast feedback, this feature is much awaited :)

- I didn't defer `file.Close()` because there's only one code path that might return the outer function - the call `os.OpenFile()` - which itself doesn't open the file if there's an error (the error is always of type `PathError`). But I agree that it's safer to defer nonetheless

- I'm not sure about `io.Copy()`. Is there a reason why we would want to retain the output string in `stdoutBuffer` when writing to a writer? Perhaps a better way would be to invoke `WriteTo()` instead?

```go
	if !cmd.OutPath.Exists() {
		_, err := stdoutBuffer.WriteTo(os.Stdout)
		if err != nil {
			return fmt.Errorf("failed to write to os.Stdout: %s", err)
		}
	} else {
		path := cmd.OutPath.Value()
		file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
		if err != nil {
			return err
		}
		defer file.Close()

		_, err = stdoutBuffer.WriteTo(file)
	}
```

I checked that `fmt.Fprint*` also uses `WriteTo()` under the hood but if we don't want to format the written content then perhaps better invoke it directly.

corylanou replied:
io.Copy will call read, effectively draining the buffer the same as calling Write (or WriteTo).  For me, it's important that we stay in `bytes`, vs fmt will convert those bytes to string and back to bytes, which is inefficient.  Not that efficiency matters a whole lot in the scheme of performance for most of what we are doing here, but any time I can avoid the `fmt` package I try to.

I try to stick to the `defer` pattern, as when code changes after the fact, it may create another return path, and inadvertently not close the file at that point.  There is no performance hit for using defer, and therefor no disadvantage to using it.  As such, I reflexively defer without even thinking about it (I don't think I even realized I changed it until you called it out).

I think your changes work just as well.  Thanks for doing that.

Kristina-Pianykh replied:
Thank you for an elaborate explanation, this makes perfect sense. There's no need not to use an optimized solution. I didn't think about the double conversion from bytes to string and back to bytes. And the `defer` pattern is definitely a good habit :)

Kristina-Pianykh commented:
Not entirely sure what causes the tests to fail. Running the tests locally works for me (classic "works on my machine").

corylanou reviewed (commented):
I had to fix some failures on the main branch yesterday due to the go version changing (and I fixed a few things in the make file).  Do you mind rebasing and seeing if you can get your tests to pass?

I pulled your branch locally and rebased but still had failing tests.

This is what I see if I run the tests after rebasing:

```
❯ make test
go test -timeout 10s -count 1 -race -cover $(go list ./... | grep -v /docs/)
--- FAIL: Test_Output_Flag_Success (2.39s)
    cli_integration_test.go:139: 
                Error Trace:    /Users/corylanou/projects/gopherguides/hype/cli_integration_test.go:139
                Error:          Received unexpected error:
                                exit status 1
                Test:           Test_Output_Flag_Success
FAIL
coverage: 73.2% of statements
FAIL    github.com/gopherguides/hype    3.390s
ok      github.com/gopherguides/hype/atomx      1.814s  coverage: 52.9% of statements
ok      github.com/gopherguides/hype/binding    1.440s  coverage: 62.3% of statements
        github.com/gopherguides/hype/cmd/hype           coverage: 0.0% of statements
input: ../
cleaned: /Users/corylanou/projects/gopherguides/hype/cmd/hype

input: ./../../readme.md
cleaned: /Users/corylanou/projects/gopherguides/hype/cmd/readme.md

input: /usr/bin/readme.md
cleaned: /usr/bin/readme.md

input: /
cleaned: /

input: /../
cleaned: /

input: ../nonexistent/
cleaned: /Users/corylanou/projects/gopherguides/hype/cmd/hype/nonexistent

--- FAIL: TestBadOutputFilePaths (0.00s)
    output_test.go:80: 
                Error Trace:    /Users/corylanou/projects/gopherguides/hype/cmd/hype/cli/output_test.go:80
                Error:          An error is expected but got nil.
                Test:           TestBadOutputFilePaths
FAIL
coverage: 30.0% of statements
FAIL    github.com/gopherguides/hype/cmd/hype/cli       0.408s
ok      github.com/gopherguides/hype/internal/lone      1.956s  coverage: 54.9% of statements
ok      github.com/gopherguides/hype/mdx        2.351s  coverage: 95.8% of statements
        github.com/gopherguides/hype/slides             coverage: 0.0% of statements
FAIL
make: *** [test] Error 1
```

Kristina-Pianykh commented:
> I had to fix some failures on the main branch yesterday due to the go version changing (and I fixed a few things in the make file). Do you mind rebasing and seeing if you can get your tests to pass?
> 
> I pulled your branch locally and rebased but still had failing tests.
> 
> This is what I see if I run the tests after rebasing:
> 
> ```
> ❯ make test
> go test -timeout 10s -count 1 -race -cover $(go list ./... | grep -v /docs/)
> --- FAIL: Test_Output_Flag_Success (2.39s)
>     cli_integration_test.go:139: 
>                 Error Trace:    /Users/corylanou/projects/gopherguides/hype/cli_integration_test.go:139
>                 Error:          Received unexpected error:
>                                 exit status 1
>                 Test:           Test_Output_Flag_Success
> FAIL
> coverage: 73.2% of statements
> FAIL    github.com/gopherguides/hype    3.390s
> ok      github.com/gopherguides/hype/atomx      1.814s  coverage: 52.9% of statements
> ok      github.com/gopherguides/hype/binding    1.440s  coverage: 62.3% of statements
>         github.com/gopherguides/hype/cmd/hype           coverage: 0.0% of statements
> input: ../
> cleaned: /Users/corylanou/projects/gopherguides/hype/cmd/hype
> 
> input: ./../../readme.md
> cleaned: /Users/corylanou/projects/gopherguides/hype/cmd/readme.md
> 
> input: /usr/bin/readme.md
> cleaned: /usr/bin/readme.md
> 
> input: /
> cleaned: /
> 
> input: /../
> cleaned: /
> 
> input: ../nonexistent/
> cleaned: /Users/corylanou/projects/gopherguides/hype/cmd/hype/nonexistent
> 
> --- FAIL: TestBadOutputFilePaths (0.00s)
>     output_test.go:80: 
>                 Error Trace:    /Users/corylanou/projects/gopherguides/hype/cmd/hype/cli/output_test.go:80
>                 Error:          An error is expected but got nil.
>                 Test:           TestBadOutputFilePaths
> FAIL
> coverage: 30.0% of statements
> FAIL    github.com/gopherguides/hype/cmd/hype/cli       0.408s
> ok      github.com/gopherguides/hype/internal/lone      1.956s  coverage: 54.9% of statements
> ok      github.com/gopherguides/hype/mdx        2.351s  coverage: 95.8% of statements
>         github.com/gopherguides/hype/slides             coverage: 0.0% of statements
> FAIL
> make: *** [test] Error 1
> ```

I've finally identified the error despite the fact the error output changed from run to run (and that the tests still magically passed on my machine).
The changing error output was caused by tests in `cli_integration_test.go` that ran in parallel and each called `os.Chdir()`. This affected the other tests running in parallel and sharing state. Changing directory conflicts with the idea of parallel tests. This is actually [nicely documented](https://cs.opensource.google/go/go/+/refs/tags/go1.24.2:src/testing/testing.go;l=1530), I should just read the friendly manual more :) https://github.com/gopherguides/hype/pull/35/commits/1ee0e7c37f58457a51f1192dda493978900590c0 ensures the tests changing working directory run sequentially. Replacing `os.Chdir()` with `t.Chdir()` also ensures proper cleanup after each test.
